modules/social_features/social_follow_taxonomy/modules/social_follow_tag/src/Plugin/Block/FollowTagBlock.php

```diff
@@ -0,0 +1,229 @@
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('entity_type.manager'),
+      $container->get('form_builder'),
+      $container->get('social_tagging.tag_service'),
+      $container->get('flag'),
+      $container->get('renderer'),
+      $container->get('flag.link_builder'),
+      $container->get('current_user')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function build() {
+    $build = [];
+    $identifiers = [];
+    $terms = [];
+
+    if ($this->tagService->allowSplit()) {
+      foreach ($this->tagService->getCategories() as $tid => $value) {
+        if (!empty($this->tagService->getChildren($tid))) {
+          $identifiers[] = social_tagging_to_machine_name($value);
+        }
+      }
+    }
+
+    foreach ($identifiers as $identifier) {
+      if (isset($_GET[$identifier])) {
+        $query = $_GET[$identifier];
+        $terms[] = $query;
+      }
+    }
+
+    $tags = [];
+    foreach ($terms as $term_ids) {
+      foreach ($term_ids as $term_key => $term) {
+
+        /** @var \Drupal\taxonomy\Entity\Term $taxonomy_term */
+        $taxonomy_term = Term::load($term);
+
+        $nodes = $this->entityTypeManager
+          ->getStorage('node')
+          ->loadByProperties(['social_tagging' => $term]);
+        $related_content = [];
+
+        foreach ($nodes as $node) {
+          $related_content[$node->bundle()]['label'] = $node->type->entity->label();
+          if ($related_content[$node->bundle()]) {
+            if (isset($related_content[$node->bundle()]['count'])) {
+              $related_content[$node->bundle()]['count'] += 1;
+            }
+            else {
+              $related_content[$node->bundle()]['count'] = 1;
+            }
+
+            $related_content[$node->bundle()]['nid'][] = $node->id();
+          }
+        }
+
+        $flag_link = $this->flagLinkBuilder->build($taxonomy_term->getEntityTypeId(), $taxonomy_term->id(), 'follow_term');
+
+        $follow = FALSE;
+        $flag = Flag::load('follow_term');
+        if ($flag instanceof FlagInterface) {
+          if (!empty($this->flagService->getFlagging($flag, $taxonomy_term, $this->currentUser))) {
+            $follow = TRUE;
+          }
+        }
+
+        if ($follow) {
+          $tags[$term] = [
+            'name' => $taxonomy_term->getName(),
+            'flag' => $flag_link,
+            'related_content' => $related_content,
+          ];
+        }
+      }
+    }
+
+    $renderable = [
+      '#theme' => 'search_follow_tag',
+      '#tagstitle' => t('Tags'),
+      '#tags' => $tags,
+    ];
+
+    $build['content']['#markup'] = $this->renderer->render($renderable);
+
+    return $build;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getCacheMaxAge() {
```

Comment:
We need to handle caching in a better way.
Please check if we can use `getCacheContexts()` or `getCacheTags()` here.